### PR TITLE
[FLINK-11827][table-runtime-blink] Introduce DataFormatConverters to convert internal data format and java format

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Introduce DataFormatConverters java class to convert internal data format and java format

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java [hunk 1]
```diff
@@ -49,6 +52,28 @@ public static int calculateHeaderInBytes(int numFields) {
...
+		} else if (type.equals(InternalTypes.FLOAT)) {
+			return 4;
+		} else if (type.equals(InternalTypes.CHAR)) {
+			return 2;
+		} else if (type.equals(InternalTypes.DATE)) {
+			return 4;
+		} else if (type.equals(InternalTypes.TIME)) {
+			return 4;
+		} else {
+			return 8;
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java]: Why the string type or map type's size is 8?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 2]
```diff
@@ -0,0 +1,1022 @@
...
+import java.util.TimeZone;
+
+import scala.Product;
+
+/**
+ * Converters between internal data format and java format.
+ */
+public class DataFormatConverters {
+
+	private static final Map<TypeInformation, DataFormatConverter> TYPE_INFO_TO_CONVERTER;
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: Why we use TypeInformation to mapping the converters? If you want to convert from internal data format to java format, i think the only type you have is `InternalType`, right?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 3]
```diff
@@ -0,0 +1,1022 @@
...
+import java.sql.Timestamp;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+
+import scala.Product;
+
+/**
+ * Converters between internal data format and java format.
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: I think more comments are needed here. In what circumstances we need convert internal format to java format, and  when do it backwards.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 4]
```diff
@@ -0,0 +1,1022 @@
...
+		} else if (typeInfo instanceof RowTypeInfo) {
+			return new RowConverter((RowTypeInfo) typeInfo);
+		} else if (typeInfo instanceof PojoTypeInfo) {
+			return new PojoConverter((PojoTypeInfo) typeInfo);
+		} else if (typeInfo instanceof TupleTypeInfo) {
+			return new TupleConverter((TupleTypeInfo) typeInfo);
+		} else if (typeInfo instanceof TupleTypeInfoBase && Product.class.isAssignableFrom(typeInfo.getTypeClass())) {
+			return new CaseClassConverter((TupleTypeInfoBase) typeInfo);
+		} else if (typeInfo instanceof BinaryArrayTypeInfo) {
+			return BinaryArrayConverter.INSTANCE;
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: Why do we need to convert between `BinaryArray` and `BinaryArrary`, same question to `BinaryMap` and `BinaryString`

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java]: It store the length and offset of variable-length part when type is string and map.

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java]: Method should change name to calculateFixLengthPartSize and add some comments.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: DataFormatConverter is oriented to Java format, while InternalType has lost its specific Java format. Only Type Information retains all its Java format information.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java (hunk 4 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: Some scenarios Udx needs to use BinaryString for performance, or put BinaryString in Row. These converters can handle these scenarios very well.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 5]
```diff
@@ -0,0 +1,1022 @@
...
+	}
+
+	public static long toLong(java.util.Date v, TimeZone timeZone) {
+		long time = v.getTime();
+		return time + (long) timeZone.getOffset(time);
+	}
+
+	/** Converts the internal representation of a SQL DATE (int) to the Java
+	 * type used for UDF parameters ({@link java.sql.Date}). */
+	public static java.sql.Date internalToDate(int v) {
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: Could we move the date time relative utilities to a specific place, such as `org.apache.flink.table.runtime.functions.DateTimeUtils`.

Currently, it is a bit messy putting date time functions in various classes.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 6]
```diff
@@ -0,0 +1,979 @@
...
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import scala.Product;
+
+/**
+ * Converters between internal data format and java format.
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: I think this is not resolved.

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: I have added comment to DataFormatConverter... I will move up

**Comment by JingsongLi** [comment]
Hi @KurtYoung @wuchong  , thanks for reviewing. I addressed your comments.

**Comment by KurtYoung** [comment]
LGTM, merging...

**Pull request closed without merging**